src/audio/tdfb/tdfb_direction.c

```diff
@@ -0,0 +1,559 @@
+
+/* Threshold for notifying user space, no more often than every 100 ms */
+#define CONTROL_UPDATE_MIN_TIME	Q_CONVERT_FLOAT(100e-3, 16)
+
+/* Emphasis filters for sound direction (IIR) */
+
+uint32_t iir_emphasis_48k[20] = {
+	0x00000002, 0x00000002, 0x00000000, 0x00000000,
+	0x00000000, 0x00000000, 0xc8cf47b5, 0x7689916a,
+	0x1dc95968, 0xc46d4d30, 0x1dc95968, 0x00000000,
+	0x00004000, 0xe16f20ea, 0x51e57f66, 0x01966267,
+	0x032cc4ce, 0x01966267, 0xfffffffe, 0x00004222
+};
+
+uint32_t iir_emphasis_16k[20] = {
+	0x00000002, 0x00000002, 0x00000000, 0x00000000,
+	0x00000000, 0x00000000, 0xd6f418ae, 0x63e7b85c,
+	0x19ae069b, 0xcca3f2ca, 0x19ae069b, 0x00000000,
+	0x00004000, 0xf504f334, 0x00000000, 0x09651419,
+	0x12ca2831, 0x09651419, 0xfffffffe, 0x0000414c
+};
+
+/**
+ * \brief This function copies data to sound direction estimation
+ *
+ * Function to copy data from component source to cross correlation buffer. The copy
+ * operation includes emphasis filtering. The copy is skipped if tracking is not
+ * enabled.
+ *
+ * \param[in,out]	cd		Component data
+ * \param[in]		ch_count	Number of channels in audio stream
+ * \param[in,out]	ch		Current channel for emphasis, incremented
+ *					and reset to zero after channel_count -1.
+ * \param[in]		x		Input PCM sample
+ */
+void tdfb_direction_copy_emphasis(struct tdfb_comp_data *cd, int ch_count, int *ch, int32_t x)
+{
+	int32_t y;
+
+	if (!cd->direction_updates)
+		return;
+
+	y = iir_df2t(&cd->direction.emphasis[*ch], x);
+	*cd->direction.wp = sat_int16(Q_SHIFT_RND(y, 31, 18)); /* 18 dB boost after high-pass */
+	cd->direction.wp++;
+	tdfb_cinc_s16(&cd->direction.wp, cd->direction.d_end, cd->direction.d_size);
+	(*ch)++;
+	if (*ch == ch_count)
+		*ch = 0;
+}
+
+/* Use function y = sqrt_int16(x), where x is Q4.12 unsigned, y is Q4.12
+ * TODO: Add scaling of input to Q4.12 with a suitable N^2 or (1/N)^2 value
+ * to increase sqrt() computation precision and avoid MIN() function to
+ * saturate for large distances (over 4m). The current virtual source
+ * distance 3m allows max 1m size array for this to work correctly without
+ * hitting the saturation.
+ */
+static inline int16_t tdfb_mic_distance_sqrt(int32_t x)
+{
+	int32_t xs;
+
+	xs = Q_SHIFT_RND(x, 24, 12);
+	xs = MIN(xs, UINT16_MAX);
+	return sqrt_int16((uint16_t)xs);
+}
+
+static int16_t max_mic_distance(struct tdfb_comp_data *cd)
+{
+	int16_t d;
+	int16_t dx;
+	int16_t dy;
+	int16_t dz;
+	int32_t d2;
+	int32_t d2_max = 0;
+	int i;
+	int j;
+
+	/* Max lag based on largest array dimension. Microphone coordinates are Q4.12 meters */
+	for (i = 0; i < cd->config->num_mic_locations; i++) {
+		for (j = 0; i < cd->config->num_mic_locations; i++) {
+			if (j == i)
+				continue;
+
+			dx = cd->mic_locations[i].x - cd->mic_locations[j].x;
+			dy = cd->mic_locations[i].y - cd->mic_locations[j].y;
+			dz = cd->mic_locations[i].z - cd->mic_locations[j].z;
+
+			/* d2 is Q8.24 meters */
+			d2 = dx * dx + dy * dy + dz * dz;
+			if (d2 > d2_max)
```

Comment:
@cujomalainey it is often my proposal not to use `a = max(a, b)` because for me that decodes into an equivalent of `if (a < b) a = b; else a = a;` where the second branch is of course redundant. Presumably, any reasonably current compiler should optimise that away, but that's just a "should." So, on hot paths, inside a nested loop, maybe it's good to have some such simple things explicit

Reply:
@lyakh fair point